File: crates/solobase/src/blocks/userportal.rs
use wafer_run::block::{Block, BlockInfo};
use wafer_run::context::Context;
use wafer_run::types::*;
use wafer_run::helpers::*;
use wafer_core::clients::config;

pub struct UserPortalBlock;

#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
impl Block for UserPortalBlock {
    fn info(&self) -> BlockInfo {
        BlockInfo {
            name: "@solobase/userportal".to_string(),
            version: "1.0.0".to_string(),
            interface: "http.handler".to_string(),
            summary: "User portal configuration endpoint".to_string(),
            instance_mode: InstanceMode::Singleton,
            allowed_modes: vec![InstanceMode::Singleton],
            admin_ui: None,
            runtime: wafer_run::types::BlockRuntime::Native,
            requires: Vec::new(),
        }
    }

    async fn handle(&self, ctx: &dyn Context, msg: &mut Message) -> Result_ {
        // GET /ext/userportal/config -> static config
        let config_val = serde_json::json!({
            "logo_url": config::get_default(ctx, "LOGO_URL", "/logo.png").await,
            "app_name": config::get_default(ctx, "APP_NAME", "Solobase").await,
            "primary_color": config::get_default(ctx, "PRIMARY_COLOR", "#6366f1").await,
            "enable_oauth": config::get_default(ctx, "ENABLE_OAUTH", "false").await,
            "allow_signup": config::get_default(ctx, "ALLOW_SIGNUP", "true").await,
            "show_powered_by": true,
            "features": {
                "files": true,
                "products": true,
                "legal_pages": true,
                "monitoring": true
            }
        });
        json_respond(msg, &config_val)
    }

    async fn lifecycle(&self, _ctx: &dyn Context, _event: LifecycleEvent) -> std::result::Result<(), WaferError> {
        Ok(())
    }
}

File: crates/solobase-cloudflare/worker/blocks-ts/userportal.ts
// TypeScript-native userportal block handler.
// Returns portal configuration (branding, feature flags) by reading config values.

import type { Message, BlockResult } from '../types';
import type { RuntimeHost } from '../host';

export async function handle(msg: Message, host: RuntimeHost): Promise<BlockResult> {
  const config = {
    logo_url: await getConfig(host, 'LOGO_URL', '/logo.png'),
    app_name: await getConfig(host, 'APP_NAME', 'Solobase'),
    primary_color: await getConfig(host, 'PRIMARY_COLOR', '#6366f1'),
    enable_oauth: await getConfig(host, 'ENABLE_OAUTH', 'false'),
    allow_signup: await getConfig(host, 'ALLOW_SIGNUP', 'true'),
    show_powered_by: true,
    features: {
      files: await getConfig(host, 'FEATURE_FILES', 'true'),
      products: await getConfig(host, 'FEATURE_PRODUCTS', 'true'),
      user_products: await getConfig(host, 'FEATURE_USER_PRODUCTS', 'true'),
      legal_pages: await getConfig(host, 'FEATURE_LEGAL_PAGES', 'true'),
      monitoring: await getConfig(host, 'FEATURE_MONITORING', 'true'),
      deployments: await getConfig(host, 'FEATURE_DEPLOYMENTS', 'true'),
    },
  };

  return {
    action: 'respond',
    response: {
      data: new TextEncoder().encode(JSON.stringify(config)),
      meta: [{ key: 'resp.content_type', value: 'application/json' }],
    },
    message: msg,
  };
}

async function getConfig(host: RuntimeHost, key: string, defaultVal: string): Promise<string> {
  const result = await host.callBlock('wafer-run/config', {
    kind: 'config.get',
    data: new TextEncoder().encode(JSON.stringify({ key })),
    meta: [],
  });
  if (result.action !== 'respond' || !result.response) return defaultVal;
  try {
    const parsed = JSON.parse(new TextDecoder().decode(result.response.data));
    return parsed?.value ?? defaultVal;
  } catch {
    return defaultVal;
  }
}
